proxy/config.go

```diff
@@ -52,27 +48,20 @@ func (c *Config) GetService(host string) *Service {
 	return nil
 }
 
-//NewConfig gets the new config.
-func NewConfig() *Config {
-	return &Config{
-		Port:       "2000",
-		Domain:     ".dev",
-		URLPattern: `.*\.dev$`,
-		Services:   nil,
-	}
-}
+//LoadServices loads the services from filepath, returns an error
+//if the configuration could not be parsed
+func (c *Config) LoadServices() error {
+	var mutex = &sync.Mutex{}
 
-//NewService gets the new service.
-func NewService(name, url string) Service {
-	return Service{
-		Name: name,
-		URL:  url,
-	}
+	mutex.Lock()
+	defer mutex.Unlock()
```

Comment:
How come you reverted to mutexes and removed channels?

Reply:
If you initiate a different mutex on every call to LoadServices, then the lock will not block correctly. You should use the same mutex on each lock. This way, the mutex is a local scop variable and it you reenter the same method again, you will reinitiate another mutex. You should declare the mutex at the package level or at the Config structure level.